### PR TITLE
Dxp cloud service stack version 4 docs

### PR DESCRIPTION
**Pull request by Alec-Shay**
Here's the current draft of the docs for DXP Cloud Stack version 4.x. I know time is limited but if there is an issue you need me to address, let me know and I'll get to it first thing next week.

Still left to-do:
- Jenkinsfile documentation (currently just a brief snippet, intending to point to a new, separate doc on this)
- Pointing to Liferay workspace documentation (once we have it)

Thanks!

### CODE REVIEW

**Comment by jrhoun** [comment]
'Just

File: docs/dxp-cloud/latest/en/reference.rst [hunk 1]
```diff
@@ -9,6 +9,8 @@ Reference
    reference/defining-environment-variables.md
    reference/support-access.md
    reference/tracking-dxp-cloud-status-and-getting-help.md
+   reference/understanding-the-dxp-cloud-project-changes-in-version-4.md
+   reference/upgrading-your-dxp-cloud-stack.md
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference.rst:13]: I'd swap the ordering...

```
   reference/upgrading-your-dxp-cloud-stack.md
   reference/understanding-the-dxp-cloud-project-changes-in-version-4.md
```

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
+# Upgrading Your DXP Cloud Stack
+
+Upgrading your repository to use the new structure in DXP Cloud version 4 allows you to take advantage of the Liferay workspace within your repository, keep your services up to date, and use a range of other features.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:3]: > take advantage of **the** Liferay workspace

Liferay Workspace (I think is maybe how it should be referred to) along with a cross reference to https://learn.liferay.com/dxp-7.x/developing-applications/tooling/liferay-workspace.html

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+Upgrading to the new repository structure involves downloading an [upgrade script](https://github.com/LiferayCloud/stack-upgrade/archive/release.zip) specifically for this purpose.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:5]: > involves **downloading** an

maybe "using" or "the use of"

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+## Preparing to Upgrade
+
+Before performing the upgrade, check to ensure you are using version 3.x of the DXP Cloud stack. Check this by looking at the version of any of your services.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:9]: In order to upgrade to version 4.x of the DXP Cloud stack, your current services must be at 3.x. To verify your current version, locate the `gradle.properties` file.

(Can you give a very specific example? I know it sounds obvious but - like, here's we're looking at the gradle properties file located in `foo/uat/liferay/gradle.properties`)

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+If you are using version 3.x, then you can check the `gradle.properties` file at the root of your repository for any of your Docker images:
+
+```properties
+liferay.workspace.lcp.liferay.image=liferaycloud/liferay-dxp:7.2.10-sp1-fp4-3.0.19
+```
+
+> The number at the end of the Docker image name indicates the version of the DXP Cloud stack you are using. Ensure that this number is some version 3.x.x (in this example, 3.0.19).
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:17]: Update to admonition syntax please.

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+Next, make sure your Git repository is on a clean branch with no uncommitted files. The upgrade script will clean uncommitted files and check out a new branch, so you must ensure all necessary files are committed.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:19]: does it do a git clean? can we explicitly say that?

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+```
+
+Perform the following steps to upgrade:
+
+1. Download the [upgrade script](https://github.com/LiferayCloud/stack-upgrade/archive/release.zip), and extract it into the root of your local DXP Cloud repository:
+
+    ```bash
+       curl https://github.com/LiferayCloud/stack-upgrade/archive/release.zip -O
+       unzip -d /path/to/project/folder/ release.zip
+    ```
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:34]: can we split the commands up into separate lines so they can be copied and pasted?

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+   Before you upgrade, make sure your repository is on a clean branch with no uncommitted files. Otherwise, these files may be deleted during the script execution.
+```
+
+Perform the following steps to upgrade:
+
+1. Download the [upgrade script](https://github.com/LiferayCloud/stack-upgrade/archive/release.zip), and extract it into the root of your local DXP Cloud repository:
+
+    ```bash
+       curl https://github.com/LiferayCloud/stack-upgrade/archive/release.zip -O
+       unzip -d /path/to/project/folder/ release.zip
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:33]: is the space intentional?

> `       unzip -d /path/to/project/folder/ release.zip`

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+```warning::
+   Before you upgrade, make sure your repository is on a clean branch with no uncommitted files. Otherwise, these files may be deleted during the script execution.
+```
+
+Perform the following steps to upgrade:
+
+1. Download the [upgrade script](https://github.com/LiferayCloud/stack-upgrade/archive/release.zip), and extract it into the root of your local DXP Cloud repository:
+
+    ```bash
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:31]: all the bash commands have one indentation too many.

image

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    ```
+
+1. Read the script's starting output, and enter your project's ID at the prompt:
+
+    ```bash
+       After the upgrade has run, you can completely undo and rerun it with the following commands:
+
+	    git checkout <original-branch-name> && git reset --hard && git branch -D upgrade-workspace; ./upgrade-workspace.sh
+
+       Press enter to continue:
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:50]: I'm not sure if all the other output from the script is helpful or noise that distracts from identifying and describing the main prompt someone is to respond to.

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+       Please enter a comma-delimited list of the different environments in your project, apart from the "common" environment.
+       For example, you can write "dev,prd". This script will only copy files from these environments and the common environment.
+       Taking the webserver service as an example, if you enter "dev", files will be copied from lcp/webserver/config/dev to webserver/configs/dev/conf.d.
+       However, files in lcp/webserver/config/anotherenv would be ignored and deleted.
+
+       Please enter a comma-delimited list of environments: dev,uat,prd
+    ```
+
+    ```note::
+       Only enter the names of Liferay environments that you add files to in your project repository. Do not add the infrastructure (`infra`) environment.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:77]: > Do not add the infrastructure (`infra`) environment

Sounds important enough to bold the "Do not add.." part.

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    ```
+
+1. If you want to install elasticsearch plugins to all of your Liferay environments, enter a comma-delimited list of their names at the prompt. Otherwise, press enter.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:80]: > elasticsearch 
capitalization

> enter a comma-delimited list of their names at the prompt. 
Which prompt?

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+The script creates and checks out a new branch in your repository, with the upgrade script added to your `.gitignore` file.
+
+Your repository is now reorganized into a structure where the `liferay` folder is now a Liferay workspace, and your services are upgraded to 4.x versions.
+
+## Next Steps
+
+After your local repository has changed, you may want to [deploy the new service versions](../build-and-deploy/walking-through-the-deployment-life-cycle.md) to a development environment.
+
+You may also want to explore some of the new functionality in the new versions of your DXP Cloud services. See the [explanation of the changes](./understanding-the-dxp-cloud-stack-changes-in-version-4.md) to learn more.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:90]: I think a verification step is probably warranted or desirable here. How do I know it worked?

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+**Contents:**
+
+* [Changes to Docker Image Definitions](#changes-to-docker-image-definitions)
+* [General Organizational Changes](#general-organizational-changes)
+* [Liferay Service Changes](#liferay-service-changes)
+* [Search Service Changes](#search-service-changes)
+* [CI Service Changes](#ci-service-changes) 
+* [Webserver Service Changes](#webserver-service-changes)
+* [Backup Service Changes](#backup-service-changes)
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:13]: The ordering and presentation of these topics should match how the intro paragraph describes them (or at least the first few)

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+Docker images for your services in version 4.x of DXP Cloud are no longer defined in the project's `gradle.properties` file. These are now defined in the `LCP.json` for each service.
+
+Together with the upgrade to the repository's structure, each service's Docker image version is also upgraded from a `3.x.x` version to a `4.x.x` version.
+
+## General Organizational Changes
+
+The biggest change to the repository is that each service's files have been moved to a folder at the root of the repository. The `lcp` folder is removed, so it is no longer in the path for any of these services. These folders themselves have been reorganized to resemble a Liferay workspace structure. <!-- TODO: Point to workspace documentation -->
+
+Several other files previously at the root of the repository (including `gradle.properties`, `build.gradle`, and `settings.gradle`) have also been moved into the `liferay` service.
+
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:26]: Old Folder Structure -> New Folder structure diagram would be helpful.

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+## Liferay Service Changes
+
+The `liferay` service folder is a Liferay workspace in version 4.x. <!-- TODO: Point to workspace documentation -->
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:29]: > The `liferay` service folder is a Liferay workspace in version 4.x.

The `liferay` service folder is now follows the functional structure of a Liferay Workspace in version 4.x. (feels more precise?)

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+| --- | --- | --- |
+| Files for deployment | lcp/liferay/deploy/{ENV}/ | liferay/configs/{ENV}/deploy/ |
+| OSGi configuration files (.cfg or .config) | lcp/liferay/config/{ENV}/ | liferay/configs/{ENV}/osgi/configs/ |
+| Other configuration overrides | lcp/liferay/config/{ENV}/ | liferay/configs/{ENV}/ |
+| Custom shell scripts | lcp/liferay/script/{ENV}/ | liferay/configs/{ENV}/scripts/ |
+| Hotfixes and patching tools | lcp/liferay/hotfix/{ENV}/ | liferay/configs/{ENV}/patching/ |
+| Licenses | lcp/liferay/license/{ENV}/ | lcp/configs/{ENV}/deploy/ |
+
+```note::
+   Files within the configs/{ENV}/ directory will be copied as overrides into the LIFERAY_HOME directory in the Liferay container in DXP Cloud.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:45]: Should use monospacing for terms like `LIFERAY_HOME`

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+```note::
+   Files in search/configs/{ENV}/ will be copied as overrides into usr/shared/elasticsearch/ in the Search container in DXP Cloud. Thus, configurations in search/configs/{ENV}/config/ such as elasticsearch.yml will be copied into usr/shared/elasticsearch/config/, overriding the default.
+```
+
+### Elasticsearch Plugins
+
+Elasticsearch plugins may be installed to your `search` service. To see the installed Elasticsearch plugins, use the shell within the `search` service and run the following command:
+
+```bash
+   bin/elasticsearch-plugin list
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:71]: too much indentation

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+All configurations within the `search` service now belong in an environment-specific `configs` directory. See the following table for the new organization of your `search` service configuration:
+
+| **Files** | **Location in 3.x** | **Location in 4.x** |
+| --- | --- | --- |
+| Elasticsearch configurations | lcp/search/config/{ENV}/ | search/configs/{ENV}/config/ |
+| Custom shell scripts | lcp/search/script/{ENV}/ | search/configs/{ENV}/scripts/ |
+| Elasticsearch license (.json) files | lcp/search/license/{ENV}/ | search/configs/{ENV}/license/ |
+
+```note::
+   Files in search/configs/{ENV}/ will be copied as overrides into usr/shared/elasticsearch/ in the Search container in DXP Cloud. Thus, configurations in search/configs/{ENV}/config/ such as elasticsearch.yml will be copied into usr/shared/elasticsearch/config/, overriding the default.
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:63]: Everytime I see a file path in an admonition its very hard to read/follow - I'm not sure if that's because they aren't formatted monospaced, or because it's a hard thing to explain in the context.

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+   Files in /webserver/configs/{ENV}/ will be copied as overrides into /etc/nginx/ in the Webserver container in DXP Cloud. Files in /webserver/configs/{ENV}/public/ will be copied as overrides into var/www/html/.
+```
+
+### Webserver Configuration Overrides
+
+You can customize the root location for the `webserver` service by adding a `liferay.conf` file into `webserver/configs/{ENV}/conf.d/`. This will override the default `liferay.conf` available in the `webserver` service image's container. Access the shell in the DXP Cloud Console to see the default `liferay.conf` file as a reference when customizing the root location.
+
+```warning::
+   Do not customize the root location using a file name other than liferay.conf, so that this file overrides the default liferay.conf and avoids an error. Other file names are instead used to define additional locations for your webserver.
+```
```
Comment (jrhoun) [inline docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:104]: I don't understand the warning. Don't change the root location of (something) by using a conf file that is named anything other than `liferay.conf`. If I do...I will get an error? Can this get clarified?

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md (hunk 4 above)
Comment (Alec-Shay) [reply docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:9]: @jrhoun I get the rephrasing from the first part of this.

As for the specific example -- you mean something not like what is in the next couple lines?

> ```properties
>liferay.workspace.lcp.liferay.image=liferaycloud/liferay-dxp:7.2.10-sp1-fp4-3.0.19
>```

Or do you just mean clarifying what the example is talking about? (I'm not sure how one would clarify the path further than what's there currently since, pre-upgrade, it would be just at the root of the repository)

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md (hunk 8 above)
Comment (Alec-Shay) [reply docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:33]: Yes, it's `unzip -d <path> <zip-name>

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md (hunk 10 above)
Comment (Alec-Shay) [reply docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:50]: I figure having a little context before the main prompt helps ground where it's supposed to appear... though having several lines there may be distracting. I can trim it a bit.

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md (hunk 4 above)
Comment (jrhoun) [reply docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:9]: Maybe a quick reminder on what the 3.x repo structure looks like? then as a reminder to ground someone on where they are?

- [folder]
- [folder]
- [gradle.propeties]
- [blah]

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md (hunk 15 above)
Comment (Alec-Shay) [reply docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:26]: @jrhoun Not sure I can imagine what that kind of diagram would look like. First thing that comes to mind is something like our folder structure diagrams:

```
folder
|__ file
|__ file
(etc.)
```

... But with how many things moved, that sort of thing I imagine would be pretty gigantic and/or cluttered (especially since there'd have to be two separate such diagrams, I'd imagine), so I'm not sure if it would still remain helpful? I mean, maybe you could simplify parts of it (grouping together similar files?), but at the cost of reduced clarity of what each thing that moved actually is. Unless maybe there's a simpler way that I'm just not seeing right away.

File: docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md (hunk 19 above)
Comment (Alec-Shay) [reply docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:63]: I'll have to search for examples of monospace in the admonitions.
I'll fix that, maybe look at it again and see if it's still hard to follow with that?

Comment (jrhoun) [reply docs/dxp-cloud/latest/en/reference/understanding-the-dxp-cloud-project-changes-in-version-4.md:63]: monospace in admonitions just requires double backticks ``blah``

File: docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md (hunk 6 above)
Comment (Alec-Shay) [reply docs/dxp-cloud/latest/en/reference/upgrading-your-dxp-cloud-stack.md:19]: I looked into this, and tested to confirm; apparently, it doesn't actually clean these files (as I thought it did). It actually automatically commits them as it's working. Updating this to reflect that.

**Comment by Alec-Shay** [comment]
See the new PR with changes: https://github.com/jrhoun/liferay-learn/pull/426
